src/dropzone/DropzoneDirective.ts

```diff
@@ -24,15 +24,40 @@ export class DropzoneDirective implements ng.IDirective {
     element: ng.IAugmentedJQuery,
     attrs: ng.IAttributes
   ) {
+    let uploadToS3 = (ctrl, file, formData) => {
+      // append AWS upload key to the form data
+      // needed in order to have a valid POST
+      let filePath = file.name;
+
+      if (file.hasOwnProperty('fullPath')) {
+        filePath = file.fullPath;
+      }
+
+      let key = ctrl.folder
+          .getPath()
+          .clone()
+          .push(filePath)
+          .toString();
+
+      formData.append('key', key);
+      return;
+    };
+
     let dropzoneConfig = {
       url: scope.$ctrl.config.url
     };
 
+    // in order to allow access to 'scope' inside the dropzone
+    // handler functions, the functions need to be wrapped in
+    // scope.$apply.
     let eventHandlers = {
       'sending': (file, xhr, formData) => {
-        // append AWS upload key to the form data
-        // needed in order to have a valid POST
-        formData.append('key', file.name);
+        scope.$apply(() => {
+          // s3 file upload
+          if (scope.$ctrl.backend.hasOwnProperty('s3')) {
```

Comment:
Don't do this. First, all backend-specific code should be *in* the backend class itself. Second, I think conditional logic around class attributes is a bad way to accomplish this, especially since the `s3` attribute is private. What if I change the API later?

Reply:
I've added a new commit where I move the 'uploadToS3' function found in the dropzone directive into AwsS3Backend.

Reply:
The new commit has been rebased into this one.